samza-kv-remote/src/main/java/org/apache/samza/table/remote/couchbase/BaseCouchbaseTableFunction.java

```diff
@@ -0,0 +1,277 @@
+  }
+
+  /**
+   * Helper method to initialize {@link Bucket}.
+   */
+  @Override
+  public void init(Context context) {
+    bucket = COUCHBASE_BUCKET_REGISTRY.getBucket(bucketName, clusterNodes, environmentConfigs);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void close() {
+    COUCHBASE_BUCKET_REGISTRY.closeBucket(bucketName, clusterNodes);
+  }
+
+  /**
+   * Check whether the exception is caused by one of the temporary failure exceptions, which are
+   * likely to be retriable.
+   * @param exception exception thrown by the table provider
+   * @return a boolean
+   */
+  public boolean isRetriable(Throwable exception) {
+    while (exception != null && !(exception instanceof TemporaryFailureException)
+        && !(exception instanceof TemporaryLockFailureException)) {
+      exception = exception.getCause();
+    }
+    return exception != null;
+  }
+
+  /**
+   * Set the timeout limit on the read / write operations. Default value is Duration.ZERO, which means no timeout.
+   * See <a href="https://docs.couchbase.com/java-sdk/2.7/client-settings.html#timeout-options"></a>.
+   * @param timeout Timeout duration
+   * @param <T> type of this instance
+   * @return Self
+   */
+  public <T extends BaseCouchbaseTableFunction<V>> T withTimeout(Duration timeout) {
+    this.timeout = timeout;
+    return (T) this;
+  }
+
+  /**
+   * Set the TTL for the data writen to Couchbase. Default value Duration.ZERO means no TTL, data will be stored forever.
+   * See <a href="https://docs.couchbase.com/java-sdk/2.7/core-operations.html#expiry"></a>.
+   * @param ttl TTL duration
+   * @param <T> type of this instance
+   * @return Self
+   */
+  public <T extends BaseCouchbaseTableFunction<V>> T withTtl(Duration ttl) {
+    this.ttl = ttl;
+    return (T) this;
+  }
+
+  /**
+   * Serde is used to serialize and deserialize values to/from byte array. If value type is not
+   * {@link com.couchbase.client.java.document.json.JsonObject}, a Serde must be provided.
+   * @param valueSerde value serde
+   * @param <T> type of this instance
+   * @return Self
+   */
+  public <T extends BaseCouchbaseTableFunction<V>> T withSerde(Serde<V> valueSerde) {
+    this.valueSerde = valueSerde;
+    return (T) this;
+  }
+
+  /**
+   * Enable role-based authentication with username and password. Note that role-based and certificate-based
+   * authentications can not be used together.
+   * @param username username
+   * @param password password
+   * @param <T> type of this instance
+   * @return Self
+   */
+  public <T extends BaseCouchbaseTableFunction<V>> T withUsernameAndPassword(String username, String password) {
+    if (environmentConfigs.sslEnabled != null && environmentConfigs.sslEnabled) {
+      throw new IllegalArgumentException(
+          "Role-Based Access Control and Certificate-Based Authentication cannot be used together.");
+    }
+    environmentConfigs.username = username;
```

Comment:
I considered about it twice. We can add check for username. But we shouldn't enforce the password to be not empty because there could be some buckets migrated from pre 5.0 Couchbase versions that has no password. In that case, simple an username without password should be able to access the bucket.